File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int low = 1, high = nums.size(); 
        int duplicate = -1; 
        while(low <= high){
            int mid = (low + high) / 2; 
            int count = 0; 
            for(int& n : nums){
                if(n <= mid) count++;  
            }

            if(count <= mid){ //정상적인 숫자일 때 
                low = mid + 1;  
            } else{
                duplicate = mid; 
                high = mid - 1; 
            }
        }

        return duplicate; 

    }
};

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int left = 1; 
        int high = 0; 
        int duplicate = -1; 
        for(int n : nums) high = Math.max(n,high); 

        while(left < high){
            int mid = left + (high - left) / 2; 
            int cnt = 0; 

            for(int n : nums){
                if(n <= mid) cnt++; 
            }

            if(cnt > mid){
                duplicate = mid; 
                high = mid - 1; 
            } else{
                left = mid + 1; 
            }
        }

        return duplicate; 
    }
}